### PR TITLE
Adding error handling to the taglib.  Synchronizing access to the glo…

### PR DESCRIPTION
**Pull request by ctoestreich**
…bal contextHolder.  Additional error handling in doFilter method.  Bumping version to 3.0.4.  Adding more tests.  Code cleanup for java warnings on doc.

https://github.com/grails/grails-core/issues/9976
https://github.com/grails-plugins/grails-cache/issues/35

### CODE REVIEW

**Comment by graemerocher** [comment]
Could you separate out the formatting changes from the actual changes, it makes the pull requests hard to review?

Can you explain the need to synchronise the thread local?

**Comment by ctoestreich** [comment]
I under the impression that the access to the context was still not completely safe when used in the scope of the filter bean as it is registered as a singleton.  It would appear that after more reading that may not be necessary with the TreadLocal object.  Concerning splitting up the changes to formatting, I will see what I can do, but the only files (non-test) with non-formatting changes are:

CacheTagLib.groovy
PageFragmentCachingFilter.groovy

File: grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy [hunk 1]
```diff
@@ -41,29 +43,34 @@ class CacheTagLib {
 			return
 		}
 
-		def cache = grailsCacheManager.getCache(attrs.cache ?: 'grailsBlocksCache')
-		def bodyClosure = ClassUtils.getPropertyOrFieldValue(body, 'bodyClosure')
-		def closureClass = bodyClosure.getClass()
-		def key = closureClass.getName()
-		def expired = false
+		try {
```
Comment (ctoestreich) [inline grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy]: Adding try catch here and fallback to the body on error

File: grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy [hunk 2]
```diff
@@ -82,34 +89,38 @@ class CacheTagLib {
...
 	def render = { attrs ->
 		if (!grailsCacheManager) {
-			out <<  g.render(attrs)
+			out << g.render(attrs)
 			return
 		}
 
-		//Make this empty string to save error later in grails gsp core
-		attrs.plugin = attrs.plugin ?: ''
+		try {
```
Comment (ctoestreich) [inline grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy]: Adding try catch here and falling back to attempt to render the template on error

File: grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy [hunk 3]
```diff
@@ -127,12 +138,11 @@ class CacheTagLib {
 		t.metaInfo.pageClass.name
 	}
 
-	protected cloneIfNecessary(content) {
+	protected static cloneIfNecessary(content) {
```
Comment (ctoestreich) [inline grails-app/taglib/grails/plugin/cache/CacheTagLib.groovy]: static modifier for non-side effect method for faster dispatch

File: src/main/groovy/grails/plugin/cache/web/ContentCacheParameters.java [hunk 4]
```diff
@@ -78,7 +79,7 @@ public HttpServletRequest getRequest() {
 	}
 
 	protected void initController() {
-		controllerClass = (GrailsControllerClass) GrailsWebRequest.lookupApplication().getArtefactForFeature(
+		controllerClass = (GrailsControllerClass) Holders.findApplication().getArtefactForFeature(
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/ContentCacheParameters.java]: Upgrade to the new application finder as the other was deprecated

File: src/main/groovy/grails/plugin/cache/web/PageInfo.java [hunk 5]
```diff
@@ -366,15 +365,17 @@ public boolean isModified(HttpServletRequest request) {
...
+     */
 	public boolean isMatch(HttpServletRequest request) {
 		String ifNoneMatch = request.getHeader(HttpHeaders.IF_NONE_MATCH);
 		String etag = getHeader(HttpHeaders.ETAG);
 		if (!StringUtils.hasLength(ifNoneMatch) || !StringUtils.hasLength(etag)) {
 			return false;
 		}
 
-		return ifNoneMatch == etag;
+		return ifNoneMatch.equals(etag);
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/PageInfo.java]: fixing string equality check

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 6]
```diff
@@ -175,21 +174,25 @@ public void setUrlMappingsHandlerMapping(UrlMappingsHandlerMapping urlMappingsHa
 
 	@Override
 	protected void doFilter(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws Exception {
-		// TODO need blocking cache stuff from CachingFilter
-
-		if(urlMappingsHandlerMapping != null) {
-			// initialize the state for the mapping
-			urlMappingsHandlerMapping.getHandler(request);
+		// todo: need blocking cache stuff from CachingFilter
+		try {
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: Adding try catch on get handler as this is *probably* what is causing the stack overflow in tomcat on spring boot when %% used

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 7]
```diff
@@ -175,21 +174,25 @@ public void setUrlMappingsHandlerMapping(UrlMappingsHandlerMapping urlMappingsHa
...
+			}
+		} catch (Exception e) {
+			log.error("PageFragmentCachingFilter could not get handler", e);
+			throw e;
 		}
 
-		initContext();
-
 		try {
+			initContext();
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: moved init context into try as there is no reason to not catch the bad stuff

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 8]
```diff
@@ -210,7 +213,7 @@ protected void doFilter(HttpServletRequest request, HttpServletResponse response
 
 			if (CollectionUtils.isEmpty(cacheOperations)) {
 				log.debug("No cacheable annotation found for {}:{} {}",
-						new Object[] { request.getMethod(), request.getRequestURI(), getContext() });
+						request.getMethod(), request.getRequestURI(), getContext());
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: Removed unnecessary typing for Object array

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 9]
```diff
@@ -257,15 +260,16 @@ protected void doFilter(HttpServletRequest request, HttpServletResponse response
 				}
 				update(caches, pageInfo, status, calculateKey(request));
 			}
-		}
-		finally {
+		} catch (RuntimeException runtimeException) {
+			log.error("PageFragmentCachingFilter has encountered an exception", runtimeException);
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: added logging for errors for better debugging of issues

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 10]
```diff
@@ -557,18 +567,22 @@ protected void logRequestDetails(HttpServletRequest request, ContentCacheParamet
...
-		log.debug("{}...", message);
-		log.debug("    method = {}", request.getMethod());
-		log.debug("    requestURI = {}", request.getRequestURI());
-		log.debug("    forwardURI = {}", WebUtils.getForwardURI(request));
-		if (WebUtils.isIncludeRequest(request)) {
-			log.debug("    includeURI = {}", request.getAttribute(WebUtils.INCLUDE_REQUEST_URI_ATTRIBUTE));
-		}
-		log.debug("    controller = {}", cacheParameters.getControllerName());
-		log.debug("    action = {}", cacheParameters.getActionName());
+		try {
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: Added try catch as this throws exception in test due to `InvokerHelper.invokeMethod(request, "getFormat", null))`

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 11]
```diff
@@ -681,13 +694,14 @@ protected boolean inspectAfterCacheEvicts(Collection<CacheOperationContext> evic
 	}
 
 	protected boolean inspectCacheEvicts(Collection<CacheOperationContext> evictions, boolean beforeInvocation) {
-		if (evictions.isEmpty()) {
+		if (evictions == null || evictions.isEmpty()) {
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: Null safety

File: src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java [hunk 12]
```diff
@@ -839,19 +849,18 @@ protected void writeResponse(final HttpServletResponse response, final PageInfo
...
 
 	protected void update(Collection<Cache> caches, PageInfo pageInfo, CacheStatus cacheStatus, String key) {
 		ValueWrapper element = cacheStatus == null ? null : cacheStatus.valueWrapper;
 		Object maxAge = pageInfo.getCacheDirectives().get("max-age");
-		int timeToLive = (maxAge instanceof Integer) ? ((Integer)maxAge) : (int)pageInfo.getTimeToLiveSeconds();
+		int timeToLive = (maxAge instanceof Integer) ? ((Integer) maxAge) : (int) pageInfo.getTimeToLiveSeconds();
 		for (Cache cache : caches) {
 			log.debug("Response ok. Adding to cache {} with key {} and ttl {}",
-					new Object[] { cache.getName(), key, getTimeToLive(element) });
+					cache.getName(), key, getTimeToLive(element));
```
Comment (ctoestreich) [inline src/main/groovy/grails/plugin/cache/web/filter/PageFragmentCachingFilter.java]: Removed unnecessary creation of Object array for params

**Pull request merged**